File: src/kotlin/app/adventofcode/impl/Day02.kt
package adventofcode.impl

import adventofcode.Puzzle
import adventofcode.Helper

class Day02(filename: String) : Puzzle(filename) {

    override fun solveFirst(): String? {
        val commands = input
            .map { it.split(" ") }
            .map { it[0].to(it[1].toInt()) }

        var depth = 0;
        var position = 0;
        for ((direction, amount) in commands) {
            when (direction) {
                "up" -> depth -= amount
                "down" -> depth += amount
                "forward" -> position += amount
                else -> Helper.debugMsg("Unknown direction ${direction}")
            }
        }

        Helper.debugMsg("Depth: $depth, Position: $position")
        return "${depth * position}"
    }

    override fun solveSecond(): String? {
        val commands = input
            .map { it.split(" ") }
            .map { it[0].to(it[1].toInt()) }

        var depth = 0;
        var position = 0;
        var aim = 0;

        for ((direction, amount) in commands) {
            when (direction) {
                "up" -> aim -= amount
                "down" -> aim += amount
                "forward" -> {
                    position += amount
                    depth += aim * amount
                }
                else -> Helper.debugMsg("Unknown direction ${direction}")
            }
        }

        Helper.debugMsg("Depth: $depth, Position: $position")
        return "${depth * position}"
    }
}

File: src/csharp/puzzleImpl/Day02.cs
using static csharp.Helper;

namespace csharp.puzzleImpl;

public class Day02 : IPuzzle
{
    private readonly string[] _input;

    public Day02(string filename)
    {
        _input = File.ReadAllLines(filename);
    }

    public string? FirstPuzzle()
    {
        var depth = 0;
        var position = 0;

        var commands = _input
            .Select(line => line.Split(" "))
            .Select(split => (split[0], int.Parse(split[1])))
            .ToList();

        foreach (var (direction, amount) in commands)
        {
            switch (direction)
            {
                case "up":
                    depth -= amount;
                    break;
                case "down":
                    depth += amount;
                    break;
                case "forward":
                    position += amount;
                    break;
                default:
                    DebugMsg($"Unknown direction {direction}");
                    break;
            }
        }

        return (depth * position).ToString();
    }

    public string? SecondPuzzle()
    {
        var depth = 0;
        var position = 0;
        var aim = 0;

        var commands = _input
            .Select(line => line.Split(" "))
            .Select(split => (split[0], int.Parse(split[1])))
            .ToList();

        foreach (var (direction, amount) in commands)
        {
            switch (direction)
            {
                case "up":
                    aim -= amount;
                    break;
                case "down":
                    aim += amount;
                    break;
                case "forward":
                    position += amount;
                    depth += aim * amount;
                    break;
                default:
                    DebugMsg($"Unknown direction {direction}");
                    break;
            }
        }

        return (depth * position).ToString();
    }
}